pkg/istio/external_authorizer.go

```diff
@@ -0,0 +1,283 @@
+package istio
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	iopv1alpha1 "istio.io/istio/operator/pkg/apis/istio/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/env"
+	istiov1alpha1 "maistra.io/istio-operator/api/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	maistrav1 "github.com/kuadrant/kuadrant-operator/api/external/maistra/v1"
+	maistrav2 "github.com/kuadrant/kuadrant-operator/api/external/maistra/v2"
+	kuadrantv1beta1 "github.com/kuadrant/kuadrant-operator/api/v1beta1"
+)
+
+const (
+	// (Sail) The istio CR must be named default to process GW API resources
+	istioCRName = "default"
+)
+
+func controlPlaneConfigMapName() string {
+	return env.GetString("ISTIOCONFIGMAP_NAME", "istio")
+}
+
+func controlPlaneProviderNamespace() string {
+	return env.GetString("ISTIOOPERATOR_NAMESPACE", "istio-system")
+}
+
+func controlPlaneProviderName() string {
+	return env.GetString("ISTIOOPERATOR_NAME", "istiocontrolplane")
+}
+
+func UnregisterExternalAuthorizer(ctx context.Context, cl client.Client, kObj *kuadrantv1beta1.Kuadrant) error {
+	isIstioInstalled, err := unregisterExternalAuthorizerIstio(ctx, cl, kObj)
+
+	if err == nil && !isIstioInstalled {
+		err = unregisterExternalAuthorizerOSSM(ctx, cl, kObj)
+	}
+
+	return err
+}
+
+func unregisterExternalAuthorizerIstio(ctx context.Context, cl client.Client, kObj *kuadrantv1beta1.Kuadrant) (bool, error) {
+	logger, _ := logr.FromContext(ctx)
+	configsToUpdate, err := getIstioConfigObjects(ctx, cl)
+	isIstioInstalled := configsToUpdate != nil
+
+	if !isIstioInstalled || err != nil {
+		return isIstioInstalled, err
+	}
+
+	kuadrantAuthorizer := newKuadrantAuthorizer(kObj.Namespace)
+
+	for _, config := range configsToUpdate {
+		hasAuthorizer, err := hasKuadrantAuthorizer(config, *kuadrantAuthorizer)
+		if err != nil {
+			return true, err
+		}
+		if hasAuthorizer {
+			if err = UnregisterKuadrantAuthorizer(config, kuadrantAuthorizer); err != nil {
+				return true, err
+			}
+
+			logger.Info("remove external authorizer from istio meshconfig")
+			if err = cl.Update(ctx, config.GetConfigObject()); err != nil {
+				return true, err
+			}
+		}
+	}
+	return true, nil
+}
+
+func unregisterExternalAuthorizerOSSM(ctx context.Context, cl client.Client, kObj *kuadrantv1beta1.Kuadrant) error {
```

Comment:
It looks like we're extracting too much. These OSSM functions do not belong to the `istio` package IMO.

The decision of whether to activate Istio or OSSM should continue to happen in the base controller or some abstraction from where we fork towards one path or the other.

Reply:
ok, moving back the logic to register either in istio or OSSM to the controller.

Reply:
done!